Training plan/schedule implementation in Kotlin

senpl opened the pull request:
This is simplest implementation of training plan/schudle/program. I used name program but it is to be decide how to officially called it.
Idea is like that:
We got programs (new in main side manu):
In programs we got ExercisesInProgram to store exercises into program (Difference from exersise is mainly stored rest break into database so each exercise could have seperate rest).
Finally we got program runner in which we could execute exercises stored in program.
Program runner support fast reps change and gestures of swipe to fast exercise changes.
This in basic form should solve:  
https://github.com/brodeurlv/fastnfitness/issues/82
and partly https://github.com/brodeurlv/fastnfitness/issues/119

Additionally it adds code in kotlin, so all projects now build with Kotlin.
I also added staho for debug implementation but it is not really needed, so it is purely optional.
I also deleted App.iml as it mess with Android studio default build(this is not needed but recommended).
It currently not allow to edit programs end exerciseInProgram order/edit, but this will change in next version I hope.
In any doubts, problems, errors or suggestions let me know.
In next version I would like to make program runner in exercise tab, but this would require to switch off tab swipe gestures so this is change that has to be discussed.

brodeurlv commented:
Hi Michal, 
Thank you for the commit. It's a pretty huge work you did here! It will take me some time to review (in particular in Kotlin) before accepting the merge. 
I was also working on this feature on my side so I'll see if we had a similar approach. 
I will try to do this asap and let you know if anything needs to be updated. 
Thanks for your contribution. It is greatly appreciated.

brodeurlv commented:
Hi Michal, first look the architecture seems pretty good.
However, I tried to launch the app and it seems that from a new fresh install, the app doesn't launch. 

>      Caused by: java.lang.IllegalStateException: getDatabase called recursively
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:357)
        at android.database.sqlite.SQLiteOpenHelper.getReadableDatabase(SQLiteOpenHelper.java:341)
        at com.easyfitness.DAO.DAOBase.getReadableDatabase(DAOBase.java:24)
        at com.easyfitness.DAO.DAOProgram.getRecord(DAOProgram.kt:32)
        at com.easyfitness.DAO.DAOProgram.programExists(DAOProgram.kt:27)
        at com.easyfitness.DAO.DAOProgram.addRecord(DAOProgram.kt:15)
        at com.easyfitness.DAO.DatabaseHelper.onCreate(DatabaseHelper.java:66) 

I guess you will also have the same issue in the onUpgrade of the DatabaseHelper. 

Can you have a look? Thanks.

senpl commented:
I hope that fix mentioned issue. If not let me know.

brodeurlv commented on app/src/main/res/layout/tab_program_runner.xml:
```diff
@@ -0,0 +1,605 @@
+android:fastScrollEnabled="false"
+            android:isScrollContainer="false"
+            android:scrollbars="none" -->
+        <LinearLayout
```
Hi Michal, 

I slept on it this night and here is an idea for the UI update that would required just few changes  (no big architectural work) . 
![UIProposal](https://user-images.githubusercontent.com/16635534/82137538-7b645a00-9819-11ea-854a-4bdf2eb847bc.PNG)

When the user click the [success button], then the record is added to the list and it switch to the next step of the program automatically. 
When the user click the [failed button], then the details are shown and the user can enter what he succeeded to do. 

What do you think ?
Thanks.

senpl commented:
about automatic date.
"I don't get why this is below the list. Should be before."
I do not think it is really needed. How many time you want to log not from today? (You still can add record with specific date from fonte).
I think it could be hidden in option, but this is not yet implemented. So i move it out of sight at this moment.

senpl commented:
progression on the program
This I think could be clarified with some sort of pager/progress indicator. I think green pager with empty not entered values will be enough.
Even gestures could go only for unfinished exercises. But this would require some sort of mark done on running program.
About failure/done mark. This would require additional database table. Nothing special but i like option to enter reps done with ease. Coz fail could be different some close to program, some not.

senpl commented:
"On "Run Program" tab, you added the Notes but they are nowhere to be seen elsewhere in the program. So this seem useless as is."
Idea is that when you store technical details they will be showed next time you train. I do not think they're needed in other places. But they could be added. I just want clear separation between exercises choice and exercises execution.

senpl commented:
"When the user click the [success button], then the record is added to the list and it switch to the next step of the program automatically.
When the user click the [failed button], then the details are shown and the user can enter what he succeeded to do.

What do you think"
I now implant it bit different. New rest timer turns on. After it finishes it moves to next exercise. This changes i plan for next version. But they are needed so with new static timer  i place it in new  version. This also in future could be optional.

brodeurlv commented:
> about automatic date.
> "I don't get why this is below the list. Should be before."
> I do not think it is really needed. How many time you want to log not from today? (You still can add record with specific date from fonte).
> I think it could be hidden in option, but this is not yet implemented. So i move it out of sight at this moment.

I agree with you that it should be hidden and this is the case in the Workout fragment. You can look at what is done there. This should be the same mechanism here.
When "Automatic time" is checked, then Date and Time are automatically set to "Today". If it is uncheck, then you can set the time you want.

brodeurlv commented:
> "On "Run Program" tab, you added the Notes but they are nowhere to be seen elsewhere in the program. So this seem useless as is."
> Idea is that when you store technical details they will be showed next time you train. I do not think they're needed in other places. But they could be added. I just want clear separation between exercises choice and exercises execution.

Ok so I didn't understand the workflow of the "notes". Thanks for clarifying.

senpl commented:
I hope init reworked now is ok. I changed rest timer and added static timer with sound in option. Temporarily totals reps are not presented. Maybe I will make it optional.

brodeurlv commented:
Hi Michal, 
The last set of commit fixed the initialization issue. Thanks. 
I observe a recurring crash when changing tabs. 
![untitled](https://user-images.githubusercontent.com/16635534/82259494-fc2b6f00-995b-11ea-84f1-24db32093f72.gif)
with the following traces: 
`    Process: com.easyfitness, PID: 19214
    java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
        at java.util.ArrayList.get(ArrayList.java:437)
        at com.easyfitness.programs.ProgramRunner.updateNote(ProgramRunner.kt:488)
        at com.easyfitness.programs.ProgramRunner.access$updateNote(ProgramRunner.kt:41)
        at com.easyfitness.programs.ProgramRunner$onViewCreated$3.afterTextChanged(ProgramRunner.kt:118)
        at android.widget.TextView.sendAfterTextChanged(TextView.java:10551)`

In Program Runner fragment : 

- Date and Time are still not filled in the record
- when you click twice on "Add" before the rest countdown is done, the Rest Button becomes crazy ! :)
![untitled2](https://user-images.githubusercontent.com/16635534/82261797-fc2d6e00-995f-11ea-97bf-6440e56c3baf.gif)
- the tab name is incorrect

In term of good practice, it would be preferable to complete this commit before adding new features. 
Thanks.

senpl commented:
Adding new exercises from program running still not what is expected(to rethink if this is needed in this version). Other things are mainly finished.
Known issues: in vertical plane around 16 exercises switching possible. Other will be showed in horizontal plane. Switching will work and makring. But no fast switching(which is acceptable I guess).

brodeurlv commented:
Hi Michal, thank you for the update. You proposed a lot of very interesting enhancement. That's a great source of inspiration for me.
However, it will be difficult for me to merge everything in my code directly. I will merge your pull request in a branch and merge things little by little. 
Also the fact that your code is mainly in Kotlin is definitively an issue for me as I don't know kotlin and I might not be able to maintain the code base if needed. However, I have started a training as I know it would be a good move for me, but this requires a lot of time and this is moving slowly.
Also, as I was myself working on the workout/program feature, I will complete my development and see how to conciliate both your development and mine. This feature is impacting a lot of core feature and I have updated some part of the architecture to make it more robust and easy to maintain and evolve.
If you want to work on other features of the app, please launch topic or an issue to discuss features and design. 
Thanks.

senpl commented:
If you using Android studio last option on selected file manu is convert to Kotlin and from kotlin to java. That way you could easily move to kotlin and from kotlin.

senpl commented:
I really not like that most of functionality I proposed are just not in program. This is just waste of my contribution. Most other apps use per exercise focus. From some reason you prefer to just overview (which might be ok for very simple exercise, but for exercise with technical description will simply not work, it also limit usage of pictures/video). I would rather give users option what they choose then force them to use current version. In other words version is far from what is currently implemented. Sure I could make alternative version, but at this point you limit your users to very limited implementation.
